File: lib/ts/models/src/linked-art/LinkedArtDescriptionMixin.ts
import {DataFactory} from "@paradicms/rdf";
import {Memoize} from "typescript-memoize";
import {LinkedArtLinguisticObject} from "./LinkedArtLinguisticObject";
import {LinkedArtModelMixin} from "./LinkedArtModelMixin";

const descriptionType = DataFactory.namedNode(
  "http://vocab.getty.edu/aat/300080091"
);

export abstract class LinkedArtDescriptionMixin extends LinkedArtModelMixin {
  @Memoize()
  get description(): LinkedArtLinguisticObject | null {
    for (const model of this.isReferredToBy) {
      if (
        model instanceof LinkedArtLinguisticObject &&
        model.hasType.some(type => type.equals(descriptionType))
      ) {
        return model;
      }
    }
    return null;
  }

  protected preMemoizeDescription(): void {
    this.description;
  }
}

File: lib/py/etl/paradicms_etl/models/linked_art/linked_art_description_mixin.py
from rdflib import URIRef

from paradicms_etl.models.linked_art.linked_art_linguistic_object import (
    LinkedArtLinguisticObject,
)
from paradicms_etl.models.linked_art.linked_art_model_mixin import LinkedArtModelMixin


class LinkedArtDescriptionMixin(LinkedArtModelMixin):
    __DESCRIPTION_TYPE = URIRef("http://vocab.getty.edu/aat/300080091")

    @property
    def description(self) -> LinkedArtLinguisticObject | None:
        for model in self.is_referred_to_by:
            if (
                isinstance(model, LinkedArtLinguisticObject)
                and self.__DESCRIPTION_TYPE in model.has_type
            ):
                return model
        return None
